full-stack-practice/client/src/Catalog.tsx

```diff
@@ -0,0 +1,66 @@
+import { useEffect, useState } from 'react';
+import { type Product } from './data';
+import { Link } from 'react-router-dom';
+import { toDollars } from './to-dollars';
+
+export function Catalog() {
+  const [products, setProducts] = useState<Product[]>();
+  useEffect(() => {
+    async function productCall() {
+      try {
+        const catalog = await fetch('/api/products');
+        if (!catalog.ok) throw new Error('catalog fetch failed');
+        const formattedResult = (await catalog.json()) as Product[];
+        setProducts(formattedResult);
+      } catch (error) {
+        console.error(error);
+        alert('Error. Click "OK" to reload catalog');
+        setProducts(undefined); // this will force a re-render
```

Comment:
I like the comment.

Since this only runs once on first mount, I believe products is already undefined, so I'm not sure this will cause a re-render. The best way to force the re-render is with an `isLoading` flag.